File: chapter_2_linked_lists/python/singly_linked_list.py
class SinglyLinkedList:
    """A Singly Linked List Class"""

    def __init__(self):
        """
        Constructor, instance attributes include head pointer
        and length
        """
        self.head = None
        self.length = 0

    def add_to_front(self, value):
        """
        Add a node to head of list
        """
        node = Node(value)
        if self.head is None:
            self.head = node
        else:
            node.next = self.head
            self.head = node
        self.length += 1
        return node

    def delete_node(self, index):
        """
        Deletes a node given the index
        """
        curr = self.head
        # delete head
        if index == 0:
            self.head = curr.next
            del(curr)
            return
        # delete last node
        if index == self.length - 1:
            while curr.next:
                prev = curr
                curr = curr.next
            prev.next = None
            del(curr)
            return
        while index != 0 and curr:
            prev = curr
            curr = curr.next
            index -= 1
        if not curr:
            raise IndexError
        prev.next = curr.next
        del(curr)

    def print_linked_list(self):
        """
        Print a linked list
        """
        curr = self.head
        while curr:
            print(curr.value, end="-> " if curr.next else "\n")
            curr = curr.next


class Node:
    """A Node class"""

    def __init__(self, value):
        """
        Constructor, instance attributes include value and 
        next pointer
        """
        self.value = value
        self.next = None

File: chapter_2_linked_lists/js/singlyLinkedList.js
class SinglyLinkedList {
  /* LinkedList constructor */
  constructor() {
    this.head = null
    this.length = 0
  }

  /* Adds a new node to the head of a singly linked list */
  addNode(value) {
    const node = new Node(value)
    if (!this.head) {
      this.head = node
    } else {
      node.next = this.head
      this.head = node
    }
    this.length++
    return node
  }

  /* Deletes a node at a given index */
  deleteNode(index) {
    /* delete head */
    let curr = this.head
    let prev
    if (index >= this.length || index < 0) throw new Error("index out of range")
    if (index === 0) {
      this.head = curr.next
    } else if (index === this.length - 1) {
      /* delete last node */
      while (curr.next) {
        prev = curr
        curr = curr.next
      }
      prev.next = null
    } else {
      while (index) {
        prev = curr
        index--
        curr = curr.next
      }
      prev.next = curr.next
    }
  }

  /* Prints all elements of a singly linked list */
  printList() {
    let curr = this.head
    let str = ""
    while (curr) {
      str += curr.value
      str += curr.next ? " -> " : ""
      curr = curr.next
    }
    console.log(str)
  }
}

class Node {
  /* Node constructor */
  constructor(value) {
    this.value = value
    this.next = null
  }
}

module.exports = {
  SinglyLinkedList,
  Node
}
